src/device_info.sh

```diff
@@ -607,40 +575,63 @@ function show_data()
 # local.
 function device_info_parser()
 {
-  local option="$1"
-  local remote="$2"
-  device_options['ip']="${configurations[ssh_ip]}"
-  device_options['port']="${configurations[ssh_port]}"
+  local long_options='help,vm,local,remote:'
+  local short_options='h'
 
-  if [[ -z "$option" && -n "${deploy_config[default_deploy_target]}" ]]; then
-    option='--'"${deploy_config[default_deploy_target]}"
+  options="$(kw_parse "$short_options" "$long_options" "$@")"
+  if [[ "$?" != 0 ]]; then
+    options_values['ERROR']="$(kw_parse_get_errors 'kw device' "$short_options" \
+      "$long_options" "$@")"
+    return 22 # EINVAL
   fi
-  option=${option:-'--local'}
 
-  case "$option" in
-    --vm)
-      device_options['target']="$VM_TARGET"
-      ;;
-    --local)
-      device_options['target']="$LOCAL_TARGET"
-      ;;
-    --remote)
-      if [[ -n "$remote" ]]; then
-        device_options['ip']=$(get_based_on_delimiter "$remote" ":" 1)
-        device_options['port']=$(get_based_on_delimiter "$remote" ":" 2)
-      fi
+  options_values['VM']=''
+  options_values['LOCAL']=''
+  options_values['REMOTE']=''
 
-      device_options['target']="$REMOTE_TARGET"
-      ;;
-    -h | --help)
-      device_info_help "$option"
-      exit 0
-      ;;
-    *)
-      complain "Invalid option: $option"
-      return 22 # EINVAL
-      ;;
-  esac
+  remote_parameters['REMOTE_IP']=''
+  remote_parameters['REMOTE_PORT']=''
+  remote_parameters['REMOTE_USER']=''
+
+  # Set basic default values
+  if [[ -n ${deploy_config[default_deploy_target]} ]]; then
+    local config_file_deploy_target=${deploy_config[default_deploy_target]}
+    options_values['TARGET']=${deploy_target_opt[$config_file_deploy_target]}
+  else
+    options_values['TARGET']="$REMOTE_TARGET"
+  fi
+
+  populate_remote_info ''
+  if [[ "$?" == 22 ]]; then
+    options_values['ERROR']="Invalid remote: $remote"
+    return 22 # EINVAL
+  fi
+
+  eval "set -- $options"
+  while [[ "$#" -gt 0 ]]; do
+    case "$1" in
+      --remote)
+        populate_remote_info "$2"
+        if [[ "$?" == 22 ]]; then
+          options_values['ERROR']="Invalid remote: $2"
+          return 22 # EINVAL
+        fi
+        options_values['TARGET']="$REMOTE_TARGET"
+        shift 2
+        ;;
+      --local)
+        options_values['TARGET']="$LOCAL_TARGET"
+        shift
+        ;;
+      --) # End of options, beginning of arguments
+        shift
+        ;;
+      *)
+        complain "Invalid option: ${1}"
+        exit 22 # EINVAL
```

Comment:
I think you just need to assign the invalid option to `options_values['ERROR']` and return 22.
Because of your last commit, the `complain` line is redundant and the `exit` line halts the execution without the chance to throw an error message and call `device_info_help`.

Reply:
nice catch!